include/drivers/rtio_sensor.h

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2019  Thomas Burdick <thomas.burdick@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_RTIO_SENSOR_H_
+#define ZEPHYR_INCLUDE_DRIVERS_RTIO_SENSOR_H_
+/**
+ * @brief RTIO Sensor Interface
+ * @defgroup rtio_sensor_interface RTIO Sensor Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <kernel.h>
+#include <device.h>
+#include <drivers/rtio.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file
+ * @brief Real-Time IO Sensor API
+ *
+ * Each  RTIO Sensor acts as a pollable stream of buffers with a driver that
+ * provides a way of deserializing the raw sensor data buffers into meaningful
+ * values.
+ *
+ * Each RTIO Sensor is expected to output buffers containing sample sets (such
+ * as 9 axis IMU data) in a device specific format understood by the driver.
+ * Each driver must then implement the sensor reader api and functionality
+ * to deserialize this stream of buffers into meaningful SI unit data. The
+ * caller may request different numerical formats of this data.
+ *
+ * *TODO* To start only 32bit floating point numerical formatted values are
+ * provided. Q format values with scale factors would be very useful for
+ * use with CMSIS DSP or many SoCs that do not contain FPUs. The reader
+ * *should* provide simple ways of converting integer readings into
+ * useful numerical formats in common functions.
+ */
+
+/**
+ * @private
+ */
+struct rtio_sensor_reader;
+
+/**
+ * @private
+ * @brief Get the number of sample sets in the current block
+ */
+typedef ssize_t (*rtio_sensor_reader_sample_sets_t)(const struct sensor_reader *reader);
+
+/**
+ * @private
+ * @brief Iterate to the next sample set in the block
+ *
+ * Returns 0 when complete, 1 otherwise
+ *
+ */
+typedef int (*rtio_sensor_reader_next_t)(struct rtio_sensor_reader *reader);
+
+/**
+ * @private
+ * @brief Get the value of a sample sets channel, if it exists
+ *
+ * The channel is an index into a device specified sample channel.
+ * For example, A 3 axis accelerometer may assign channel 0 to
+ * the X axis, 1 to the Y axis, and 2 to the Z axis.
+ *
+ * *TODO* discuss how the channel identifier might be obtained for something like
```

Comment:
The one downside of the above call is that you don't necessarily know what each float represents. We could instead ask this API for lets say, accelerometer X,Y,Z in that order and get it back. That would rectify needing lots of calls to get each value. The question would be what happens if that sample set has accelerometer Z but not x and y, the single get call would return an error for the non-existent values. The multi-get call I think would need a bitvector to describe that adding more complications potentially.

Timestamps might be more than just a u32_t, there may be some +/- understanding of the timestamp of some value, and a clock source based on what @pabigot mentioned. There might also be some flag specifying the source of the timestamp beyond just the clock source. Was it an estimate interpolated from the blocks begin/end timestamps? Was it the time we handled the interrupt in the mcu? The time the transfer completed? There's lots of possibilities there to contend with. Adding function calls to obtain those bits of information is entirely possible here, adding the timestamp as an out parameter removes those possibilities.

Reply:
Sounds like the timestamp pieces are still being discussed and probably belong in the core RTIO API anyway, so nevermind :).

Regarding multiple values returned:

> One is that you don't necessarily know what each float represents.

We do if we specify it by contract in the API for individual sensor event types, Android-style -- I personally find this a nice API from a user perspective.

> The question would be what happens if that sample set has accelerometer Z but not x and y, the single get call would return an error for the non-existent values. 

I know this is just an example for the sake of discussion, but I would say it sounds like that hardware couldn't implement the 3d acceleration sensor API in this case.

You could still write a "raw" driver for it, or, if there's enough people demanding it, a 2D acceleration type could be added in, no?

Reply:
So this particular example might be a poor one, but often times you can enable/disable individual channels on these sensors. For example to detect taps/drops/whatever in a low power manner. Other times they have multiple mems sensors (gyro+accel) and you can enable/disable gyro or accel individually. There are *many* scenarios like this.

Reply:
In which case the user should know what data to ignore based on their configuration, no?

If a headline RTIO feature is that it enables zero-copy, I don't get why we have to copy out the data one float at a time.

Reply:
If we're getting to floats, there's some logic regardless! Unless the sensor provides floats.

I think the general result could be fixed by using a struct param and bitflags here to signify "value exists or does not exist" which right, most times, won't matter

Reply:
Sure, but with this we're writing each float twice, not once, before it can be accessed. For composite structures there is a savings to be had if we can just point at the data in the block (not to mention all the extra branching).

Your idea seems reasonable.